keanu-project/src/main/java/io/improbable/keanu/network/ProtobufReader.java

```diff
@@ -0,0 +1,236 @@
+                Longs.toArray(value.getDoubleVal().getShapeList()));
+        }
+    }
+
+    private <T> Vertex<T> createVertexFromProtoBuf(KeanuSavedBayesNet.Vertex vertex,
+                                                   Map<KeanuSavedBayesNet.VertexID, Vertex> existingVertices) {
+        Class vertexClass;
+        try {
+            vertexClass = Class.forName(vertex.getVertexType());
+        } catch (ClassNotFoundException e) {
+            throw new IllegalArgumentException("Unknown Vertex Type Specified: " + vertex.getVertexType(), e);
+        }
+
+        Map<String, Vertex> parentsMap = getParentsMap(vertex, existingVertices);
+
+
+        return instantiateVertex(vertexClass, parentsMap, vertex.getConstantValue());
+    }
+
+    private Vertex instantiateVertex(Class vertexClass,
+                                     Map<String, Vertex> parentsMap,
+                                     KeanuSavedBayesNet.VertexValue value) {
+        Constructor<Vertex> loadConstructor = getAnnotatedConstructor(vertexClass);
+        Parameter[] parameters = loadConstructor.getParameters();
+
+        Object[] arguments = new Object[parameters.length];
+
+        for (int i = 0; i < parameters.length; i++) {
+            LoadParentVertex parentVertexAnnotation = parameters[i].getAnnotation(LoadParentVertex.class);
+            LoadVertexValue vertexValueAnnotation = parameters[i].getAnnotation(LoadVertexValue.class);
+
+            if (parentVertexAnnotation != null) {
+                arguments[i] = parentsMap.get(parentVertexAnnotation.value());
+            } else if (vertexValueAnnotation != null) {
+                arguments[i] = extractInitialValue(value);
+            }
+        }
+
+        for (Object argument: arguments) {
+            if (argument == null) {
+                throw new IllegalArgumentException("Unable to create Vertex of class: " + vertexClass + " from Value: "
+                    + value + " with parents: " + parentsMap);
+            }
+        }
+
+        try {
+            return loadConstructor.newInstance(arguments);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Failed to create new Vertex", e);
+        }
+    }
+
+    private Tensor extractInitialValue(KeanuSavedBayesNet.VertexValue value) {
+        switch (value.getValueTypeCase()) {
+            case INTVAL:
+                return extractIntValue(value);
+
+            case BOOLVAL:
+                return extractBoolValue(value);
+
+            case DOUBLEVAL:
+                return extractDoubleValue(value);
+
+            default:
+                return null;
+        }
+    }
+
+    private Constructor getAnnotatedConstructor(Class vertexClass) {
+        Constructor[] constructors = vertexClass.getConstructors();
+
+        for (Constructor constructor : constructors) {
+            Parameter[] parameters = constructor.getParameters();
+
+            if (parameters.length > 0 &&
```

Comment:
Doesn't all of the parameters need to be annotated?

Reply:
Yeah - I do verify that in my full test - I didn't know whether to redo that here though - all I wanted to do was find the one annotated constructor that we'll use to create the object.